scene/resources/material.cpp

```diff
@@ -1125,6 +1134,14 @@ uniform vec3 uv2_offset;
 	code += R"(
 void vertex() {)";
 
+	if (layers_cull_mode == LAYERS_CULL_VERTEX) {
+		code += R"(
+	if ((CAMERA_VISIBLE_LAYERS & layers) == 0u) { // Begin vertex cull mode
+		VERTEX = vec3(1.0 / 0.0);
```

Comment:
My experience with this trick to discard a triangle as been mixed at best. On at least a few devices I tested on this resulted in corrupted triangles instead of discarding the triangle

Reply:
Do you know if there is a best way to do this? I've seen several suggestions online, but i don't have the experience to know which might work the best:
* Set all the vertices to NaN/Inf
* Set all vertices outside of the view frustum
* Set all the vertices to the same value so that the primitive collapses to a point

I did make a note in the class documentation about the vertex method potentially not working on all systems. Don't know if that would be enough to allow this through:
![image](https://github.com/user-attachments/assets/26cec324-1149-4eae-92cf-84075c8289d9)

Reply:
I've had the best luck setting the vertex outside the view frustum. To do that here you would need to use the `POSITION` built in instead of `VERTEX` because `VERTEX` is the view space position. 

Overall though I'm not certain the shader is the best place for this check. With this approach the draw call still happens even if it's filtered out. 

You could pass the later into the rendering server and then filter the material when the instance buffer is being set up. That way you can skip the draw call entirely (and all the associated setup cost)

Reply:
Okay, that makes sense about stopping it before the draw call. I think both lyuma and tetrapod (in the proposal) mentioned the same thing. The main reason i went about it this way is i wasn't sure how to do it the other way :sweat_smile: (I know C++ and Godot shader code, but not OpenGL or Vulcan.) I might look into it, though, and see if i can figure it out. Would you have any pointers on where i could start, such as where visual instances are culled?

If i'm able to figure it out, would it make more sense for this property to go on `Material` instead of `BaseMaterial3D`? I suppose there's probably a significant difference between 3D and 2D layers/rendering, which might make things significantly more complicated if working with just `Material`. But maybe i could copy how `next_pass` works and only expose the `layers` property for 3D materials.

Reply:
I used `VERTEX = vec3(-INF);` to hide particles on contact back when I implemented particle collision (the current implementation sets `ACTIVE = false` instead, a particles-only built-in).